### ISSUE TITLE
C# support

### ISSUE THREAD

**Issue by KanarekLife** [github/super-linter#150]
I think adding c# support would be nice.

I think that [Resharper Command Line Tools](https://www.jetbrains.com/resharper/download/index.html#section=commandline) would be a great fit.

**Comment by programatt** [comment]
I'm giving it a shot. https://github.com/github/super-linter/pull/189

**Comment by BroderPeters** [comment]
After a few attempts on my own with inspiration from #189, I run into the following struggles.

**TLDR: Alpine seems to struggle with a few "concepts" that Resharper Command Line Tools is using.**

As I was trying to run the inspectcode script in the docker image I got the first error saying `linux-x64/mono/bin/mono-sgen: not found`. After a short research I found this related issue: https://youtrack.jetbrains.com/issue/RIDER-19133
> I believe this issue has to do with Alpine/Arch Linux not liking the prebuilt binaries (so it won't execute them).

Therefore, I tried the [workaround](https://youtrack.jetbrains.com/issue/RIDER-19133#focus=streamItem-27-4102488.0-0) but ran into the next error: 
```
Failed to fill the environment container with handler JetBrains.Application.Environment.HostParameters.DefaultHostCommandLineFromProcessRuntimeHostMixin. 
Could not load file or assembly 'PresentationCore, Version=4.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies.
```

As I'm not that experienced with alpine and the dotnet runtime dependencies I stopped for now.
If somebody has a clue how to fix this I'm happy to join there!

I'll have a look at https://github.com/SonarSource/sonar-dotnet later on as an alternative.

**Comment by programatt** [comment]
@BroderPeters sonar is a good linter as well. The biggest downside is in my experience the rules that it uses are installed with the linter and not passed in. So you wouldn't be able to allow per repo linting rules to be passed in.

**Comment by BroderPeters** [comment]
After having a look at the sonar products I don't think this will fit either. There are only IDE extension or service hosted products, which don't really make sense in a GitHub action or this projects setup itself.

Another solution to use Resharper Command Line Tools I was thinking of was to use it through the `dotnet tools` feature from the dotnet core SDK. But installing the full SDK into the docker image doesn't useful as well. And for alpine it needs a few steps to set up I think.

**Comment by Bjego** [comment]
@BroderPeters yep static code analysis has been deprecated in flavor of so called roslyn analysers. These are dependencies which you add to your project and then you only build your code as usual. And by the way SonarQubes scanner isn't static as well. It requires a code build anyway (https://docs.sonarqube.org/latest/analysis/scan/sonarscanner-for-msbuild/). 

See this blogpost: https://devblogs.microsoft.com/dotnet/write-better-code-faster-with-roslyn-analyzers/ 
And this docu page: https://docs.microsoft.com/en-us/visualstudio/code-quality/roslyn-analyzers-overview?view=vs-2019

Maybe it's better to run a recommended roslyn analyser (See other discussion this would also require a sdk)? This could get installed via command line and would run automatically in a dotnet build command.
https://blog.markvincze.com/automated-portable-code-style-checking-in-net-core-projects/ (haven't found a commandline based documentation for FxCop)

Not sure which analyser is the best one here. We are using Sonarqubes nuget package within our organisation (it works even if it's not connected to a sonarqube). 

But Microsoft recommends FxCop - This is why I would choose FxCop for Superlinter

**Comment by Bjego** [comment]
@BroderPeters not sure if you are a member of Github, or @admiralAwkbar, but maybe it's worth having a quick chat with the dotnet team for the best solution (e.g. @dsplaisted) as Github and Dotnet are parts of Microsoft ;-)

**Comment by BroderPeters** [comment]
@Bjego thanks for the detailed and reasonable answer of yours! 
Your arguments regarding the Rosalyn analyzers seem valid to me and I think it deserve a shot. I also like your steps of behavior and options.

Regarding your second comment: No, I'm not a member of GitHub (unfortunately (; ). But I do agree that if there is the possibility to ask the creators to do so.

**Comment by eddynaka** [comment]
instead of going with resharper, we could try to use [dotnet-format](https://github.com/dotnet/format)

**Comment by broder-peters-k24** [comment]
> instead of going with resharper, we could try to use [dotnet-format](https://github.com/dotnet/format)

@eddynaka I'm not sure how dotnet-format will fulfill the purpose of linting:
> dotnet-format is a code formatter for dotnet that applies style preferences to a project or solution.

At least for me linting is about programmatic stylistic and errors, which, to my current understanding, dotnet-format doesn't do.
I think what would make more sense would be to use [roslyn analyzers](https://github.com/dotnet/roslyn-analyzers).

> Roslyn analyzers analyze your code for style, quality and maintainability, design and other issues

**Comment by eddynaka** [comment]
Dotnet format reads editor config where you can set the lint properties. So, if you use fxcop and set some rules there, format will catch.

Other things like spacing, end line, etc.

**Comment by Bjego** [comment]
Think @eddynaka idea sounds good. See this blogpost as well: https://www.meziantou.net/enforce-dotnet-code-style-in-ci-with-dotnet-format.htm
